File: src/main/java/designpatterns/Subject.java
package designpatterns;

import java.util.Set;

public interface Subject {
    public void addObserver(Observer observer);

    public void removeObserver(Observer observer);

    public Set<Observer> getObservers();

    public default void updated() {
        for (Observer o : this.getObservers())
            o.update(this);
    }
}

File: src/main/java/designpatterns/Subject.kt
package designpatterns

interface Subject {

    val observers: Set<Observer>
    fun addObserver(observer: Observer) {
        this.observers.plus(observer)
    }

    fun removeObserver(observer: Observer) {
        this.observers.minus(observer)
    }

    fun updated() {
        for (o in this.observers)
            o.update(this)
    }
}
